src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs

```diff
@@ -22,30 +23,45 @@ namespace Microsoft.CodeAnalysis.LanguageServer.Handler.SemanticTokens
     internal class SemanticTokensHelpers
     {
         /// <summary>
-        /// Returns the semantic tokens data for a given document with an optional range.
+        /// Returns the semantic tokens data for a given document with an optional ranges.
         /// </summary>
         public static async Task<int[]> ComputeSemanticTokensDataAsync(
             ClientCapabilities capabilities,
             Document document,
-            LSP.Range? range,
+            LSP.Range[]? ranges,
             ClassificationOptions options,
             CancellationToken cancellationToken)
         {
-            var tokenTypesToIndex = SemanticTokensSchema.GetSchema(capabilities.HasVisualStudioLspCapability()).TokenTypeToIndex;
+            if (ranges is not null)
+            {
+                // Expecting sorted ranges within the document
+                Array.Sort(ranges, static (r1, r2) => r1.CompareTo(r2));
+            }
 
+            var tokenTypesToIndex = SemanticTokensSchema.GetSchema(capabilities.HasVisualStudioLspCapability()).TokenTypeToIndex;
             var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
             var text = await document.GetValueTextAsync(cancellationToken).ConfigureAwait(false);
+            var classifiedSpansLength = ranges?.Length ?? 1;
+            using var _ = ArrayBuilder<ClassifiedSpan>.GetInstance(classifiedSpansLength, out var classifiedSpans);
```

Comment:
>classifiedSpansLength [](http://example.com/codeflow?start=67&length=21)

This isn't a good estimate of the final size, as this is the number of spans being classified, not the number of classified tokens.